File: clearnotation_reference/normalizer.py
"""Normalizer: converts a validated parsed tree into the renderer-facing normalized AST."""

from __future__ import annotations

import unicodedata
from pathlib import Path
from typing import Any, cast

from .errors import ValidationFailure
from .inline_parser import InlineParser
from .models import (
    BlockDirective,
    BlockNode,
    BlockQuote,
    CodeSpan,
    Comment,
    Document,
    Emphasis,
    Heading,
    InlineDirective,
    InlineNode,
    Link,
    ListItem,
    NBlockQuote,
    NCallout,
    NExtensionBlock,
    NFigure,
    NHeading,
    NListItem,
    NMathBlock,
    NOrderedItem,
    NOrderedList,
    NParagraph,
    NRef,
    NSourceBlock,
    NTable,
    NTableCell,
    NTableRow,
    NThematicBreak,
    NToc,
    NUnorderedList,
    NormalizedBlock,
    NormalizedDocument,
    NormalizedInline,
    Note,
    OrderedList,
    Paragraph,
    SourceBlock,
    Strong,
    Text,
    ThematicBreak,
    UnorderedList,
)
from .registry import Registry
from .utils import split_table_row


class Normalizer:
    def __init__(self, registry: Registry) -> None:
        self.registry = registry
        self.note_counter = 1
        self.notes: list[Note] = []
        self.slug_counts: dict[str, int] = {}
        self._source_path: Path | None = None
        self._config: dict[str, Any] | None = None
        self._include_stack: frozenset[Path] = frozenset()
        self._include_depth: int = 0

    def normalize(
        self,
        document: Document,
        *,
        source_path: Path | None = None,
        config: dict[str, Any] | None = None,
        _include_stack: frozenset[Path] | None = None,
        _include_depth: int = 0,
    ) -> NormalizedDocument:
        self._source_path = source_path or document.path
        self._config = config
        self._include_stack = _include_stack if _include_stack is not None else frozenset()
        self._include_depth = _include_depth
        blocks = self._normalize_blocks(document.blocks, pending_anchor=None)
        return NormalizedDocument(
            meta=dict(document.meta),
            blocks=blocks,
            notes=list(self.notes),
        )

    def _normalize_blocks(
        self,
        blocks: list[BlockNode],
        pending_anchor: str | None,
    ) -> list[NormalizedBlock]:
        result: list[NormalizedBlock] = []
        for block in blocks:
            if isinstance(block, Comment):
                continue

            if isinstance(block, Heading):
                heading_id, pending_anchor = self._heading_id(block, pending_anchor)
                result.append(NHeading(
                    level=block.level,
                    id=heading_id,
                    content=self._normalize_inlines(block.children),
                ))
                continue

            if isinstance(block, Paragraph):
                block_id, pending_anchor = self._consume_anchor(pending_anchor)
                result.append(NParagraph(
                    content=self._normalize_inlines(block.children),
                    id=block_id,
                ))
                continue

            if isinstance(block, ThematicBreak):
                result.append(NThematicBreak())
                continue

            if isinstance(block, BlockQuote):
                block_id, pending_anchor = self._consume_anchor(pending_anchor)
                result.append(NBlockQuote(
                    lines=[self._normalize_inlines(line) for line in block.lines],
                    id=block_id,
                ))
                continue

            if isinstance(block, UnorderedList):
                block_id, pending_anchor = self._consume_anchor(pending_anchor)
                result.append(NUnorderedList(
                    items=[
                        NListItem(
                            content=self._normalize_inlines(item.children),
                            blocks=self._normalize_blocks(item.blocks, pending_anchor=None),
                        )
                        for item in block.items
                    ],
                    id=block_id,
                ))
                continue

            if isinstance(block, OrderedList):
                block_id, pending_anchor = self._consume_anchor(pending_anchor)
                result.append(NOrderedList(
                    items=[
                        NOrderedItem(
                            ordinal=item.ordinal,
                            content=self._normalize_inlines(item.children),
                            blocks=self._normalize_blocks(item.blocks, pending_anchor=None),
                        )
                        for item in block.items
                    ],
                    id=block_id,
                ))
                continue

            if isinstance(block, SourceBlock):
                block_id, pending_anchor = self._consume_anchor(pending_anchor)
                result.append(NSourceBlock(
                    language=block.language,
                    text=block.text,
                    id=block_id,
                ))
                continue

            if isinstance(block, BlockDirective):
                if block.name == "include" and self._source_path is not None and self._config is not None:
                    inlined, pending_anchor = self._inline_include(
                        block, pending_anchor,
                    )
                    result.extend(inlined)
                    continue
                normalized, pending_anchor = self._normalize_directive(
                    block, pending_anchor,
                )
                if normalized is not None:
                    result.append(normalized)
                continue

        return result

    def _normalize_directive(
        self,
        block: BlockDirective,
        pending_anchor: str | None,
    ) -> tuple[NormalizedBlock | None, str | None]:
        if block.name == "anchor":
            anchor_id = cast(str, block.attrs.get("id"))
            return None, anchor_id

        if block.name == "include":
            # When source_path/config are available, inlining is handled in
            # _normalize_blocks via _inline_include.  This fallback keeps the
            # v0.1 behaviour for callers that don't provide those arguments.
            return None, pending_anchor

        if block.name == "toc":
            block_id, pending_anchor = self._consume_anchor(pending_anchor)
            return NToc(id=block_id), pending_anchor

        if block.name == "callout":
            block_id, pending_anchor = self._consume_anchor(pending_anchor)
            return NCallout(
                kind=cast(str, block.attrs.get("kind", "")),
                title=block.attrs.get("title"),
                compact=bool(block.attrs.get("compact", False)),
                blocks=self._normalize_blocks(block.blocks, pending_anchor=None),
                id=block_id,
            ), pending_anchor

        if block.name == "figure":
            block_id, pending_anchor = self._consume_anchor(pending_anchor)
            return NFigure(
                src=cast(str, block.attrs.get("src", "")),
                blocks=self._normalize_blocks(block.blocks, pending_anchor=None),
                id=block_id,
            ), pending_anchor

        if block.name == "math":
            block_id, pending_anchor = self._consume_anchor(pending_anchor)
            return NMathBlock(
                text=block.raw_text,
                id=block_id,
            ), pending_anchor

        if block.name == "table":
            block_id, pending_anchor = self._consume_anchor(pending_anchor)
            return self._normalize_table(block, block_id), pending_anchor

        if block.name == "source":
            block_id, pending_anchor = self._consume_anchor(pending_anchor)
            return NSourceBlock(
                language=cast(str, block.attrs.get("language", "")),
                text=block.raw_text,
                id=block_id,
            ), pending_anchor

        # Unknown extension directive: wrap as NExtensionBlock
        block_id, pending_anchor = self._consume_anchor(pending_anchor)
        return NExtensionBlock(
            type_name=block.name,
            data=dict(block.attrs),
            blocks=self._normalize_blocks(block.blocks, pending_anchor=None),
            id=block_id,
        ), pending_anchor

    def _inline_include(
        self,
        block: BlockDirective,
        pending_anchor: str | None,
    ) -> tuple[list[NormalizedBlock], str | None]:
        """Resolve an ``::include`` directive by recursively normalizing the target."""
        from .parser import ReferenceParser
        from .validator import ReferenceValidator
        from .config import load_config

        src = cast(str, block.attrs.get("src"))
        assert self._source_path is not None
        target = (self._source_path.parent / src).resolve()

        # Circular detection: check if target is already in the include chain
        if target in self._include_stack:
            chain = " -> ".join(str(p.name) for p in self._include_stack)
            raise ValidationFailure(
                "circular_include",
                f"Circular include detected: {chain} -> {target.name}",
                line=block.source_line,
            )

        # Depth check
        if self._include_depth >= 10:
            raise ValidationFailure(
                "include_depth_exceeded",
                f"Include depth exceeds maximum of 10 levels",
                line=block.source_line,
            )

        # Read, parse, validate the target file
        source_text = target.read_text(encoding="utf-8")
        parser = ReferenceParser(self.registry)
        doc = parser.parse_document(source_text, target)

        config, _ = load_config(target)
        validator = ReferenceValidator(self.registry)
        validator.validate(doc, config=config)

        # Normalize recursively with shared slug/note state
        child = Normalizer(self.registry)
        child.slug_counts = self.slug_counts
        child.note_counter = self.note_counter
        child_doc = child.normalize(
            doc,
            source_path=target,
            config=self._config,
            _include_stack=self._include_stack | {target},
            _include_depth=self._include_depth + 1,
        )

        # Absorb shared state back
        self.note_counter = child.note_counter
        self.notes.extend(child_doc.notes)

        return child_doc.blocks, pending_anchor

    def _normalize_table(
        self,
        block: BlockDirective,
        block_id: str | None,
    ) -> NTable:
        raw_rows = [
            split_table_row(line)
            for line in block.raw_text.splitlines()
            if line.strip()
        ]
        rows: list[NTableRow] = []
        for raw_row in raw_rows:
            cells: list[NTableCell] = []
            for cell_text in raw_row:
                parsed = InlineParser(cell_text, self.registry, line=block.source_line).parse()
                cells.append(NTableCell(content=self._normalize_inlines(parsed)))
            rows.append(NTableRow(cells=cells))

        return NTable(
            header=bool(block.attrs.get("header", False)),
            align=block.attrs.get("align"),
            rows=rows,
            id=block_id,
        )

    def _normalize_inlines(self, inlines: list[InlineNode]) -> list[NormalizedInline]:
        result: list[NormalizedInline] = []
        for node in inlines:
            if isinstance(node, Text):
                result.append(node)
                continue
            if isinstance(node, CodeSpan):
                result.append(node)
                continue
            if isinstance(node, Strong):
                result.append(Strong(children=self._normalize_inlines(node.children)))
                continue
            if isinstance(node, Emphasis):
                result.append(Emphasis(children=self._normalize_inlines(node.children)))
                continue
            if isinstance(node, Link):
                result.append(Link(
                    label=self._normalize_inlines(node.label),
                    target=node.target,
                ))
                continue
            if isinstance(node, Note):
                note = Note(
                    children=self._normalize_inlines(node.children),
                    number=self.note_counter,
                )
                self.note_counter += 1
                self.notes.append(note)
                result.append(note)
                continue
            if isinstance(node, InlineDirective):
                if node.name == "ref":
                    target = cast(str, node.attrs.get("target", ""))
                    result.append(NRef(target=target))
                    continue
                # Other inline directives not handled in v0.1 core
                continue
        return result

    def _heading_id(
        self,
        block: Heading,
        pending_anchor: str | None,
    ) -> tuple[str, str | None]:
        if pending_anchor is not None:
            return pending_anchor, None
        base = self._slugify(self._plain_text(block.children))
        count = self.slug_counts.get(base, 0) + 1
        self.slug_counts[base] = count
        slug = base if count == 1 else f"{base}-{count}"
        return slug, None

    def _consume_anchor(
        self,
        pending_anchor: str | None,
    ) -> tuple[str | None, str | None]:
        if pending_anchor is not None:
            return pending_anchor, None
        return None, None

    def _plain_text(self, inlines: list[InlineNode]) -> str:
        parts: list[str] = []
        for node in inlines:
            if isinstance(node, Text):
                parts.append(node.value)
            elif isinstance(node, CodeSpan):
                parts.append(node.value)
            elif isinstance(node, (Strong, Emphasis)):
                parts.append(self._plain_text(node.children))
            elif isinstance(node, Link):
                parts.append(self._plain_text(node.label))
        return "".join(parts)

    def _slugify(self, text: str) -> str:
        normalized = unicodedata.normalize("NFKD", text).lower()
        chars: list[str] = []
        last_dash = False
        for ch in normalized:
            if ch.isascii() and ch.isalnum():
                chars.append(ch)
                last_dash = False
                continue
            if not last_dash:
                chars.append("-")
                last_dash = True
        return "".join(chars).strip("-")

File: clearnotation-js/src/normalizer.ts
/**
 * Normalizer: converts BlockNote BNBlock[] into the renderer-facing
 * NormalizedDocument. This is the JS counterpart of the Python normalizer.py.
 *
 * The core challenge is converting flat BNStyledText (with style flags like
 * clnStrong: true) back into nested NormalizedInline trees (NStrong with
 * children).
 */

// These types mirror the editor's converter types. Defined locally so this
// package compiles independently without reaching outside its rootDir.

/** A styled text span (mirrors editor BNStyledText). */
interface BNStyledText {
  type: "text";
  text: string;
  styles: Record<string, boolean | string>;
}

/** A link containing styled text spans (mirrors editor BNLink). */
interface BNLink {
  type: "link";
  href: string;
  content: BNStyledText[];
}

/** Inline content: either styled text or a link. */
type BNInlineContent = BNStyledText | BNLink;

/**
 * A generic block in the BlockNote document model (mirrors editor BNBlock).
 */
export interface BNBlock {
  id?: string;
  type: string;
  props: Record<string, string | number | boolean>;
  content: BNInlineContent[];
  children: BNBlock[];
  parseError?: boolean;
}
import type {
  NormalizedBlock,
  NormalizedDocument,
  NormalizedInline,
  NText,
  NCodeSpan,
  NStrong,
  NEmphasis,
  NLink,
  NNote,
  NRef,
  NHeading,
  NParagraph,
  NThematicBreak,
  NBlockQuote,
  NListItem,
  NUnorderedList,
  NOrderedList,
  NOrderedItem,
  NToc,
  NCallout,
  NFigure,
  NMathBlock,
  NTable,
  NTableRow,
  NTableCell,
  NSourceBlock,
} from "./types";
import { slugify } from "./utils";

// ---------------------------------------------------------------------------
// State container
// ---------------------------------------------------------------------------

interface NormalizerState {
  noteCounter: number;
  notes: NNote[];
  slugCounts: Map<string, number>;
  meta: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Normalize a BlockNote block array into a NormalizedDocument.
 */
export function normalize(
  blocks: BNBlock[],
  meta?: Record<string, unknown>,
): NormalizedDocument {
  const state: NormalizerState = {
    noteCounter: 1,
    notes: [],
    slugCounts: new Map(),
    meta: meta ? { ...meta } : {},
  };

  const normalized = normalizeBlocks(blocks, state, undefined);

  return {
    meta: state.meta,
    blocks: normalized,
    notes: [...state.notes],
  };
}

// ---------------------------------------------------------------------------
// Block normalization
// ---------------------------------------------------------------------------

function normalizeBlocks(
  blocks: BNBlock[],
  state: NormalizerState,
  pendingAnchor: string | undefined,
): NormalizedBlock[] {
  const result: NormalizedBlock[] = [];

  for (const blk of blocks) {
    switch (blk.type) {
      case "clnHeading": {
        const level = (blk.props.level as number) || 1;
        let headingId: string;
        if (pendingAnchor !== undefined) {
          headingId = pendingAnchor;
          pendingAnchor = undefined;
        } else {
          const base = slugify(plainText(blk.content));
          const count = (state.slugCounts.get(base) ?? 0) + 1;
          state.slugCounts.set(base, count);
          headingId = count === 1 ? base : `${base}-${count}`;
        }
        result.push({
          type: "heading",
          level,
          id: headingId,
          content: normalizeInlines(blk.content, state),
        } satisfies NHeading);
        break;
      }

      case "clnParagraph": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const para: NParagraph = {
          type: "paragraph",
          content: normalizeInlines(blk.content, state),
        };
        if (blockId !== undefined) para.id = blockId;
        result.push(para);
        break;
      }

      case "clnThematicBreak": {
        result.push({ type: "thematic_break" } satisfies NThematicBreak);
        break;
      }

      case "clnCodeBlock": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const src: NSourceBlock = {
          type: "source_block",
          language: (blk.props.language as string) || "",
          text: (blk.props.code as string) || "",
        };
        if (blockId !== undefined) src.id = blockId;
        result.push(src);
        break;
      }

      case "clnBlockquote": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        // Split content by newline text nodes into lines
        const lines = splitIntoLines(blk.content, state);
        const bq: NBlockQuote = { type: "blockquote", lines };
        if (blockId !== undefined) bq.id = blockId;
        result.push(bq);
        break;
      }

      case "clnUnorderedList": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const listItem: NListItem = {
          content: normalizeInlines(blk.content, state),
          blocks: normalizeBlocks(blk.children, state, undefined),
        };
        const ul: NUnorderedList = {
          type: "unordered_list",
          items: [listItem],
        };
        if (blockId !== undefined) ul.id = blockId;
        result.push(ul);
        break;
      }

      case "clnOrderedList": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const ordinal = (blk.props.startNumber as number) || 1;
        const ol: NOrderedList = {
          type: "ordered_list",
          items: [
            {
              ordinal,
              content: normalizeInlines(blk.content, state),
              blocks: normalizeBlocks(blk.children, state, undefined),
            },
          ],
        };
        if (blockId !== undefined) ol.id = blockId;
        result.push(ol);
        break;
      }

      case "clnToc": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const toc: NToc = { type: "toc" };
        if (blockId !== undefined) toc.id = blockId;
        result.push(toc);
        break;
      }

      case "clnCallout": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const callout: NCallout = {
          type: "callout",
          kind: (blk.props.kind as string) || "",
          title: (blk.props.title as string) || undefined,
          compact: Boolean(blk.props.compact),
          blocks: normalizeBlocks(blk.children, state, undefined),
        };
        if (blockId !== undefined) callout.id = blockId;
        result.push(callout);
        break;
      }

      case "clnFigure": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const figure: NFigure = {
          type: "figure",
          src: (blk.props.src as string) || "",
          blocks: normalizeBlocks(blk.children, state, undefined),
        };
        if (blockId !== undefined) figure.id = blockId;
        result.push(figure);
        break;
      }

      case "clnMath": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const math: NMathBlock = {
          type: "math_block",
          text: (blk.props.rawContent as string) || "",
        };
        if (blockId !== undefined) math.id = blockId;
        result.push(math);
        break;
      }

      case "clnSource": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const src: NSourceBlock = {
          type: "source_block",
          language: (blk.props.language as string) || "",
          text: (blk.props.rawContent as string) || "",
        };
        if (blockId !== undefined) src.id = blockId;
        result.push(src);
        break;
      }

      case "clnTable": {
        const blockId = pendingAnchor;
        pendingAnchor = undefined;
        const table = normalizeTable(blk, state, blockId);
        result.push(table);
        break;
      }

      case "clnMeta": {
        // Merge props into document meta; don't emit a block
        for (const [key, val] of Object.entries(blk.props)) {
          state.meta[key] = val;
        }
        break;
      }

      case "clnAnchor": {
        // Set pending anchor for next block
        pendingAnchor = (blk.props.id as string) || undefined;
        break;
      }

      case "clnInclude": {
        // Skip includes (not inlined in v0.1)
        break;
      }

      default:
        // Unknown block types are silently skipped
        break;
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// Table normalization
// ---------------------------------------------------------------------------

function normalizeTable(
  blk: BNBlock,
  state: NormalizerState,
  blockId: string | undefined,
): NTable {
  // Table data comes as a JSON string in props.tableData
  const tableData = blk.props.tableData as string | undefined;
  const rows: NTableRow[] = [];

  if (tableData) {
    try {
      const parsed = JSON.parse(tableData) as {
        rows?: Array<{ cells?: Array<{ content?: BNInlineContent[] }> }>;
      };
      if (parsed.rows) {
        for (const row of parsed.rows) {
          const cells: NTableCell[] = [];
          if (row.cells) {
            for (const cell of row.cells) {
              cells.push({
                content: normalizeInlines(cell.content || [], state),
              });
            }
          }
          rows.push({ cells });
        }
      }
    } catch {
      // Invalid JSON — produce empty table
    }
  }

  const table: NTable = {
    type: "table",
    header: Boolean(blk.props.header),
    align: blk.props.align
      ? (blk.props.align as string).split(",")
      : undefined,
    rows,
  };
  if (blockId !== undefined) table.id = blockId;
  return table;
}

// ---------------------------------------------------------------------------
// Inline normalization — flat-to-tree conversion
// ---------------------------------------------------------------------------

/**
 * Convert flat BNInlineContent[] to nested NormalizedInline[].
 *
 * The algorithm handles BNStyledText with style flags (clnStrong, clnEmphasis,
 * clnNote, clnCode, clnRef) and converts them into nested tree nodes.
 */
function normalizeInlines(
  items: BNInlineContent[],
  state: NormalizerState,
): NormalizedInline[] {
  return flatToTree(items, state);
}

/** Priority-ordered marks for the flat-to-tree grouping algorithm. */
const MARK_PRIORITY: string[] = ["clnNote", "clnStrong", "clnEmphasis"];

function flatToTree(
  items: BNInlineContent[],
  state: NormalizerState,
): NormalizedInline[] {
  const result: NormalizedInline[] = [];
  let i = 0;

  while (i < items.length) {
    const item = items[i];

    // Handle links
    if (item.type === "link") {
      const bnLink = item as BNLink;
      result.push({
        type: "link",
        label: flatToTree(bnLink.content, state),
        target: bnLink.href,
      } satisfies NLink);
      i++;
      continue;
    }

    // From here, item is BNStyledText
    const styled = item as BNStyledText;

    // clnRef → NRef (text is the target)
    if (styled.styles.clnRef) {
      result.push({ type: "ref", target: styled.text } satisfies NRef);
      i++;
      continue;
    }

    // clnCode → NCodeSpan (no nesting inside code)
    if (styled.styles.clnCode) {
      result.push({ type: "code_span", value: styled.text } satisfies NCodeSpan);
      i++;
      continue;
    }

    // Find outermost mark from priority list
    const mark = MARK_PRIORITY.find((m) => styled.styles[m]);

    if (mark === undefined) {
      // Plain text — no marks
      result.push({ type: "text", value: styled.text } satisfies NText);
      i++;
      continue;
    }

    // Group consecutive items sharing this mark
    let j = i;
    while (j < items.length) {
      const next = items[j];
      if (next.type !== "text") break;
      if (!(next as BNStyledText).styles[mark]) break;
      j++;
    }

    // Extract the group, strip the mark, recurse
    const group = items.slice(i, j) as BNStyledText[];
    const stripped: BNStyledText[] = group.map((g) => {
      const newStyles = { ...g.styles };
      delete newStyles[mark];
      return { type: "text" as const, text: g.text, styles: newStyles };
    });

    const children = flatToTree(stripped, state);

    switch (mark) {
      case "clnNote": {
        const noteNumber = state.noteCounter++;
        const note: NNote = {
          type: "note",
          children,
          number: noteNumber,
        };
        state.notes.push(note);
        result.push(note);
        break;
      }
      case "clnStrong":
        result.push({ type: "strong", children } satisfies NStrong);
        break;
      case "clnEmphasis":
        result.push({ type: "emphasis", children } satisfies NEmphasis);
        break;
    }

    i = j;
  }

  return result;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Extract plain text from BNInlineContent[] for slug generation.
 */
function plainText(items: BNInlineContent[]): string {
  const parts: string[] = [];
  for (const item of items) {
    if (item.type === "text") {
      parts.push((item as BNStyledText).text);
    } else if (item.type === "link") {
      parts.push(plainText((item as BNLink).content));
    }
  }
  return parts.join("");
}

/**
 * Split inline content into lines (for blockquotes).
 * Text nodes containing \n are split; other items go on the current line.
 */
function splitIntoLines(
  items: BNInlineContent[],
  state: NormalizerState,
): NormalizedInline[][] {
  const lines: NormalizedInline[][] = [[]];

  for (const item of items) {
    if (item.type === "text") {
      const styled = item as BNStyledText;
      const parts = styled.text.split("\n");
      for (let p = 0; p < parts.length; p++) {
        if (p > 0) lines.push([]);
        if (parts[p]) {
          const wrappedItems: BNStyledText[] = [
            { type: "text", text: parts[p], styles: styled.styles },
          ];
          const converted = flatToTree(wrappedItems, state);
          lines[lines.length - 1].push(...converted);
        }
      }
    } else {
      // Links and other items go on the current line
      const converted = flatToTree([item], state);
      lines[lines.length - 1].push(...converted);
    }
  }

  return lines;
}
